File: LeetCode/Python/Medium/1143-Longest-Common-Subsequence/solution.py
'''
Solution 1: pure recursion. This solution receives Time Limit Exceeded when run. DO NOT submit this!!!
'''
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        result=0

        n=len(text1)
        
        m=len(text2)

        i=0
        
        j=0

        result=self.helper(text1, text2, n, m, i, j)

        return result

    def helper(self, text1: str, text2: str, n: int, m: int, i: int, j: int) -> int:
        if i >= n or j >= m:
            return 0

        subproblem_solution=0

        if text1[i]==text2[j]:
            subproblem_solution=1 + self.helper(text1, text2, n, m, i + 1, j + 1)
        else:
            subproblem_solution=max(self.helper(text1, text2, n, m, i + 1, j), self.helper(text1, text2, n, m, i, j + 1))

        return subproblem_solution

'''
Solution 2: recursion + memoization
'''
class Solution:
    def __init__(self):
        self.memo=list()
    
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        result=0

        n=len(text1)
        
        m=len(text2)

        self.memo=[[-1 for _ in range(0, m)] for _ in range(0, n)]

        i=0
        
        j=0

        result=self.helper(text1, text2, n, m, i, j)

        return result

    def helper(self, text1: str, text2: str, n: int, m: int, i: int, j: int) -> int:
        if i >= n or j >= m:
            return 0

        if self.memo[i][j]!=-1:
            return self.memo[i][j]

        subproblem_solution=0

        if text1[i]==text2[j]:
            subproblem_solution=1 + self.helper(text1, text2, n, m, i + 1, j + 1)
        else:
            subproblem_solution=max(self.helper(text1, text2, n, m, i + 1, j), self.helper(text1, text2, n, m, i, j + 1))

        self.memo[i][j]=subproblem_solution
        
        return subproblem_solution

File: LeetCode/Rust/Medium/1143-Longest-Common-Subsequence/solution.rs
use std::cmp;

impl Solution
{
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32
    {
        let mut result: i32=0;

        let n: usize=text1.len();

        let m: usize=text2.len();

        let mut memo: Vec<Vec<i32>>=vec![vec![-1; m]; n];

        let text1_characters: Vec<char>=text1.chars().collect::<Vec<char>>();

        let text2_characters: Vec<char>=text2.chars().collect::<Vec<char>>();

        let mut i: usize=0;

        let mut j: usize=0;

        result=Self::helper(&text1_characters, &text2_characters, &mut memo, n, m, i, j);

        result
    }

    fn helper(text1_characters: &Vec<char>, text2_characters: &Vec<char>, memo: &mut Vec<Vec<i32>>, n: usize, m: usize, i: usize, j: usize) -> i32
    {
        if i >= n || j >= m
        {
            return 0;
        }

        if memo[i][j]!=-1
        {
            return memo[i][j];
        }

        let mut subproblem_solution: i32=0;

        if text1_characters[i]==text2_characters[j]
        {
            subproblem_solution=1 + Self::helper(text1_characters, text2_characters, memo, n, m, i + 1, j + 1);
        }
        else
        {
            subproblem_solution=std::cmp::max(Self::helper(text1_characters, text2_characters, memo, n, m, i + 1, j), Self::helper(text1_characters, text2_characters, memo, n, m, i, j + 1));
        }

        memo[i][j]=subproblem_solution;

        subproblem_solution
    }
}